recipes-app/iot2050-firmware-update/files/iot2050-firmware-update

```diff
@@ -272,6 +283,73 @@ class FirmwareUpdate(object):
         if firmware_md5_digest != firmware_flash_md5_digest:
             raise UpgradeError("Firmware digest verification failed")
 
+    def read_out_fw_from_flash(self, firmware_length=0x8c0000):
+        try:
+            if not os.path.exists(self.back_fw_path):
+                os.mkdir(self.back_fw_path)
+        except OSError:
+            raise UpgradeError("Create rollback firmware folder failure")
+
+        print("\nBackup Firmware started")
+
+        flash_md5_digest = []
+        i = 0
+        while i < 2:
+            mtd_num = 0
+            mtd_in_memory = b''
+            firmware_len = firmware_length
+
+            flash_md5 = hashlib.md5()
+            while True:
+                if firmware_len <= 0:
+                    break
+
+                mtd_dev_path, mtd_size, mtd_erasesize, mtd_name = self.get_mtd_info(
+                    mtd_num)
+
+                mtd_pos = 0
+                try:
+                    mtd_dev = os.open(mtd_dev_path, os.O_SYNC | os.O_RDWR)
+                except IOError as e:
+                    raise UpgradeError("Opening {} failed: {}".format(mtd_dev_path, e.strerror))
+
+                while mtd_pos < mtd_size and firmware_len > 0:
+                    mtd_content = os.read(mtd_dev, mtd_erasesize)
+                    mtd_pos += mtd_erasesize
+                    firmware_len -= mtd_erasesize
+                    flash_md5.update(mtd_content)
+                    mtd_in_memory += mtd_content
+
+                os.close(mtd_dev)
+
+                mtd_num += 1
+
+            flash_md5_digest.append(flash_md5.hexdigest())
+            i += 1
+
+        if flash_md5_digest[0] != flash_md5_digest[1]:
+            print("Backup Firmware falied")
+            sys.exit(1)
+
+        info = tarfile.TarInfo(name=self.backup_fw_name)
+        info.size = len(mtd_in_memory)
+        with tarfile.TarFile(self.rollback_fw_tar, 'w') as tar:
+            tar.addfile(info, io.BytesIO(mtd_in_memory))
+        print('Backup Firmware ended\n')
+
+    def rollback_firmware(self):
+        if not os.path.exists(self.rollback_fw_tar) or not tarfile.is_tarfile(self.rollback_fw_tar):
+            print('No rollback firmware exists')
+            print('It could be that you have not done any upgrades')
+            sys.exit(1)
+
+        try:
+            with tarfile.open(self.rollback_fw_tar) as f:
+                self.firmware = f.extractfile(self.backup_fw_name)
+                self.update_firmware()
+        except tarfile.ExtractError:
+            print("Extract firmware error")
```

Comment:
Extract back-up rollback firmware failed/error, so that we could tell apart from the extraction error of the normal upgrade package.